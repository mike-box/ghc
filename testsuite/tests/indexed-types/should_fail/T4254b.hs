{-# LANGUAGE TypeFamilies, FunctionalDependencies, RankNTypes, MultiParamTypeClasses #-}
module T4254b where

class FD a b | a -> b where
  op :: a -> b;
  op = undefined

instance FD Int Bool

fails :: forall a b. (a~Int,FD a b) => a -> Bool
fails  = op
-- Could fail: no proof that b~Bool
-- But can also succeed; it's not a *wanted* constraint
