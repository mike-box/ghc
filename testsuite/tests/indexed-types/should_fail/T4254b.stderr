
T4254b.hs:10:10: error:
    • Couldn't match type ‘b’ with ‘Bool’
        arising from a functional dependency between constraints:
          ‘FD Int Bool’
            arising from a type ambiguity check for
            the type signature for ‘fails’ at T4254b.hs:10:10-48
          ‘FD Int b’
            arising from the type signature for:
                           fails :: forall a b.
                                    (a ~ Int, FD a b) =>
                                    a -> Bool at T4254b.hs:10:10-48
      ‘b’ is a rigid type variable bound by
        the type signature for:
          fails :: forall a b. (a ~ Int, FD a b) => a -> Bool
        at T4254b.hs:10:10-48
    • In the ambiguity check for ‘fails’
      To defer the ambiguity check to use sites, enable AllowAmbiguousTypes
      In the type signature:
        fails :: forall a b. (a ~ Int, FD a b) => a -> Bool
